code/controllers/subsystem/SSeconomy.dm

```diff
@@ -146,9 +152,18 @@ SUBSYSTEM_DEF(economy)
 	process_job_tasks()
 	if(next_mail_delay <= world.time)
 		if(!is_admin_level(SSshuttle.supply.z) || SSshuttle.supply.areaInstance.moving)
-			return
-		next_mail_delay = 15 MINUTES + world.time
-		SSshuttle.mail_delivery()
+			next_mail_delay = 15 MINUTES + world.time
+			SSshuttle.mail_delivery()
+	var/all_completed = TRUE
+	for(var/datum/virology_goal/V in GLOB.virology_goals)
+		if(!V.completed)
+			all_completed = FALSE
+			break
+	if(all_completed)
+		GLOB.archived_virology_goals += GLOB.virology_goals
+		GLOB.virology_goals = list(new/datum/virology_goal/propertysymptom, new/datum/virology_goal/virus, new/datum/virology_goal/virus/stealth)
+		for(var/obj/machinery/computer/pandemic/P in GLOB.pandemics)
+			P.print_goal_orders()
```

Comment:
strongly dislike all of this being in `SSeconomy`

Reply:
And also, checking each `fire` if goals where completed is very bad.
You should only track completion only when it's required. Because most of the time, it will just do useless work.
And also, you should randomize virology goals, not hardcode them.
And do such algorithm:
1) We generate goals for the round
2) If goals are completed - generate new set.
You can specify `tier` in `virology_goal` datum for example, to check whether you can use goal in first generation or not.

Reply:
Absolutely agree with Gaxeer, as I said in the discord, we should check completion on receiving virus samples in cargo